src/modules/fancyzones/lib/FancyZones.cpp

```diff
@@ -566,49 +575,138 @@ LRESULT CALLBACK FancyZones::s_WndProc(HWND window, UINT message, WPARAM wparam,
 
 void FancyZones::UpdateZoneWindows() noexcept
 {
-    auto callback = [](HMONITOR monitor, HDC, RECT *, LPARAM data) -> BOOL
     {
+      std::unique_lock writeLock(m_lock);
+      m_zoneWindowMap.clear();
+    }
+
+    struct device
+    {
+      std::wstring _deviceID;
+      HMONITOR _handle;
+
+      std::optional<std::wstring_view> get_generated_part_of_id() const
+      {
+          if (const auto sharpPos = _deviceID.find(L'#'); sharpPos != std::wstring::npos)
+          {
+            return std::wstring_view{_deviceID.data() + sharpPos + 1, size(_deviceID) - sharpPos - 1};
+          }
+          return std::nullopt;
+      }
+    };
+    auto callback = [](HMONITOR monitor, HDC, RECT *, LPARAM captured_ptr) -> BOOL
+    {
+        const auto doContinueEnumeration = TRUE;
         MONITORINFOEX mi;
         mi.cbSize = sizeof(mi);
-        if (GetMonitorInfo(monitor, &mi))
+        if (GetMonitorInfo(monitor, &mi) == FALSE)
         {
-            DISPLAY_DEVICE displayDevice = { sizeof(displayDevice) };
-            PCWSTR deviceId = nullptr;
+          return doContinueEnumeration;
+        }
+        DISPLAY_DEVICE displayDevice = { sizeof(displayDevice) };
 
-            bool validMonitor = true;
-            if (EnumDisplayDevices(mi.szDevice, 0, &displayDevice, 1))
+        if (EnumDisplayDevices(mi.szDevice, 0, &displayDevice, 1))
+        {
+            if (WI_IsFlagSet(displayDevice.StateFlags, DISPLAY_DEVICE_MIRRORING_DRIVER))
             {
-                if (WI_IsFlagSet(displayDevice.StateFlags, DISPLAY_DEVICE_MIRRORING_DRIVER))
-                {
-                    validMonitor = FALSE;
-                }
-                else if (displayDevice.DeviceID[0] != L'\0')
-                {
-                    deviceId = displayDevice.DeviceID;
-                }
+                return doContinueEnumeration;
             }
+        }
+        auto devices = reinterpret_cast<std::vector<device> *>(captured_ptr);
+        device newDevice{{}, monitor};
+        if (displayDevice.DeviceID[0] != L'\0')
+        {
+            std::array<wchar_t, 256> parsedId{};
+            ParseDeviceId(displayDevice.DeviceID, parsedId.data(), size(parsedId));
+            newDevice._deviceID = parsedId.data();

```

Comment:
this looks like you are storing in `newDevice._deviceID` a pointer to a local buffer `parseId`, that will be deleted on the next line. Is this correct? If it is, a comment explaining why it is ok to do so would be great.

Reply:
@bzoz _deviceID is a `std::wstring` => has `operator=` overloaded for `wchar_t*` => deep copy happens